### PR TITLE
Implement point in time backups (snapshots)

### PR DESCRIPTION
**Pull request by WesselVS**
A solution to mitigate #52

### CODE REVIEW

File: bin/backup-standalone.sh [hunk 1]
```diff
@@ -36,6 +36,7 @@ HELP=$(
                                passed to Admin Client. Only useful if you have additional connection options
 --help                         Prints this message
 --debug                        Print Debug information (if using the environment variable, set it to 'y')
+--snapshot-mode                One-off backup mode
```
Comment (itadventurer) [inline bin/backup-standalone.sh]: I would prefer to name it just `--snapshot` – i think that makes it clear enough

File: src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  /**
+   * Obtain end offsets for each given partition
+   */
+  public Map<TopicPartition, Long> getEndOffsets(Collection<TopicPartition> partitions) {
+    Properties props = new Properties();
+    props.put("bootstrap.servers", bootstrapServers);
+    props.put("key.deserializer", ByteArrayDeserializer.class.getName());
+    props.put("value.deserializer", ByteArrayDeserializer.class.getName());
+    KafkaConsumer<Byte[], Byte[]> consumer = new KafkaConsumer<>(props);
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java]: This will not work in general. See `OffsetSource` how to create a consumer with the required configuration. E.g.: https://github.com/itadventurer/kafka-backup/blob/master/src/main/java/de/azapps/kafkabackup/common/offset/OffsetSource.java#L51

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 3]
```diff
@@ -50,15 +71,49 @@ public void start(Map<String, String> props, OffsetSink overrideOffsetSink) {
...
+     */
+    private void terminateIfCompleted() {
+        boolean terminate = true;
+        for (TopicPartition partition : endOffsets.keySet()) {
+            Long endOffset = endOffsets.get(partition);
+            Long currentOffset = currentOffsets.getOrDefault(partition, -1L);
+
+            if (currentOffset < endOffset - 1) {
+                terminate = false;
+                break;
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: I think you could even `return` here – or do you mislike doing so?

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 4]
```diff
@@ -11,23 +12,30 @@
...
-import java.util.Map;
+import java.util.*;
 
 public class BackupSinkTask extends SinkTask {
     private static final Logger log = LoggerFactory.getLogger(BackupSinkTask.class);
     private Path targetDir;
     private Map<TopicPartition, PartitionWriter> partitionWriters = new HashMap<>();
     private long maxSegmentSizeBytes;
     private OffsetSink offsetSink;
+    private SinkTaskContext sinkTaskContext;
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: why do we need this (and the other variables apart from the adminClient) as instance vars?

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 5]
```diff
@@ -11,23 +12,30 @@
...
+    private BackupSinkConfig config;
+    private AdminClient adminClient;
+    private Map<TopicPartition, Long> endOffsets;
+    private Map<TopicPartition, Long> currentOffsets = new HashMap<>();
+    private EndOffsetReader endOffsetReader;
+    private String bootstrapServers;
+    private java.util.function.Consumer<Integer> exitFunction;
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:35]: I do not know whether this is such a good idea. We are basically killing-off Kafka Connect. For the restore procedure I implemented the termination differently: https://github.com/itadventurer/kafka-backup/blob/master/bin/restore-standalone.sh#L277

Tl;dr:
* Print a specific message when the restore is finished
* The `restore-standalone.sh` is looking for the message
* `kill` Kafka Connect and give it time to terminate it gracefully

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 6]
```diff
@@ -50,15 +71,49 @@ public void start(Map<String, String> props, OffsetSink overrideOffsetSink) {
             if(overrideOffsetSink != null) {
                 offsetSink = overrideOffsetSink;
             } else {
-                AdminClient adminClient = AdminClient.create(config.adminConfig());
+                adminClient = AdminClient.create(config.adminConfig());
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: Would it make sense to use the adminClient in the EndOffsetReader instead creating an own client?

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 7]
```diff
@@ -69,15 +124,26 @@ public void put(Collection<SinkRecord> records) {
                 if (sinkRecord.kafkaOffset() % 100 == 0) {
                     log.debug("Backed up Topic {}, Partition {}, up to offset {}", sinkRecord.topic(), sinkRecord.kafkaPartition(), sinkRecord.kafkaOffset());
                 }
+                if (config.snapShotMode()) {
+                    currentOffsets.put(topicPartition, sinkRecord.kafkaOffset());
+                    log.debug("Update offset for topic {}, partition {}, offset {}", sinkRecord.topic(), sinkRecord.kafkaPartition(), sinkRecord.kafkaOffset());
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: I think this makes sense to log-debug for all `put()` calls and not just for snapshots – or do not do that at all ;)

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 8]
```diff
@@ -101,10 +168,23 @@ public void open(Collection<TopicPartition> partitions) {
                     // The offset was not found, so rather than forcibly set the offset to 0 we let the
                     // consumer decide where to start based upon standard consumer offsets (if available)
                     // or the consumer's `auto.offset.reset` configuration
+
+                    // if we are in snapshot mode, then just start at zero.
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:160]: I don't think that we want it that way – this would cause a rebuild of the whole backup in case Kafka Connect crashes. I see where this comes from I am also not absolutely happy with the current behaviour. I think we should do that in an extra PR. Lets discuss it in #100 further.

File: src/test/java/de/azapps/kafkabackup/sink/BackupSinkTaskTest.java [hunk 9]
```diff
@@ -80,6 +81,53 @@ public void simpleTest() throws Exception {
...
+        endOffsets.put(new TopicPartition(TOPIC1, 0), 13L);
+
+        records.add(new Record(TOPIC1, 0, KEY_BYTES, VALUE_BYTES, 0));
+        records.add(new Record(TOPIC1, 0, null, null, 1));
+        records.add(new Record(TOPIC1, 0, KEY_BYTES, VALUE_BYTES, 10));
+
+        List<TopicPartition> partitions = new ArrayList<>();
+        partitions.add(new TopicPartition(TOPIC1, 0));
+
+        AtomicBoolean endConditionCheck = new AtomicBoolean();
```
Comment (itadventurer) [inline src/test/java/de/azapps/kafkabackup/sink/BackupSinkTaskTest.java:104]: Cool! Did not know about that before ;)

**Review by itadventurer — CHANGES_REQUESTED** [review]
Thank you for the PR! This will be very useful for many users. I have a few comments we should discuss – but only smaller things ;)

File: bin/backup-standalone.sh (hunk 1 above)
Comment (WesselVS) [reply bin/backup-standalone.sh]: Sure! Fixed it.

File: src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java (hunk 2 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java]: Can you elaborate on this a little bit? I made some changes here such that the configuration is derived from the Sink configuration, but I needed to add the config for key/value serde manually.

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 3 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: I'm a Scala developer, 'return' is not common there :-) . Will fix.

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 4 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: I removed some instance variables, but retained most of them. I need to keep track of offsets and be able to override stuff for testing.

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 5 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:35]: I understand and totally agree that this is not ideal. However, I left this in because kafka connect seems to terminate reasonably graceful, judging by the console output. I assume they installed some JVM shutdown hook. The wait-for-stdout solution in bash is definitely not ideal either. If you prefer that solution, then I will implement that, but also consider the System.exit approach.

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 6 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: Not sure that I follow, I am not aware that the AdminClient is capable of retrieving the end offsets of partitions.

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 7 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: Removed it

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 8 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:160]: I'm not too happy about this either, but to me this seems reasonable in the case of snapshots. If kafka connect crashes, upon restart, the last offset will be retrieved from the file system, so the backup will resume. My line of thought for the zero offset is a 'new target folder' situation. The start offset from file system would be -1, but the consumer offset will not be 0. This is undesirable for a snapshot. There are other ways to fix this though.

**Comment by WesselVS** [comment]
Thanks for your review and your work on this library!

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 5 above)
Comment (itadventurer) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:35]: Ah if this works gracefully I actually prefer your approach! Have created #101 to change the behaviour during restore ;)

File: src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java (hunk 2 above)
Comment (itadventurer) [reply src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java]: I would have done it the same way ;)

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 6 above)
Comment (itadventurer) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: Sorry, you were right: this should not work with the AdminClient

File: src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    for (Map.Entry<TopicPartition, Long> partitionOffset: offsets.entrySet()) {
+      if (partitionOffset.getValue() == 0L) {
+        toRemove.add(partitionOffset.getKey()); // don't store empty offsets
+      }
+    }
+
+    for (TopicPartition partition: toRemove) {
+      offsets.remove(partition);
+    }
+    consumer.close();
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java]: Shouldn't we use `try {} finally { consumer.close() }` here?
I am not really sure how to do this nicely… maybe even make this a static function. What do you think?

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java [hunk 11]
```diff
@@ -101,10 +156,23 @@ public void open(Collection<TopicPartition> partitions) {
...
+                        context.offset(topicPartition, 0);
+                    }
+
                     log.info("Resetting offset for {} based upon existing consumer group offsets or, if "
                             + "there are none, the consumer's 'auto.offset.reset' value.", topicPartition);
                 }
+
                 this.partitionWriters.put(topicPartition, partitionWriter);
+                this.currentOffsets.put(topicPartition, lastWrittenOffset);
+                log.debug("last written offset " + lastWrittenOffset);
```
Comment (itadventurer) [inline src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: we should either print the topicpartition here or remove the message

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 8 above)
Comment (itadventurer) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java:160]: Ah sorry, I was confused by the diff view. missed the else branch. You are correct. With #100 I would even set the `context.offset(topicPartition, 0)` as the default behaviour even without the snapshotMode.

**Review by itadventurer — CHANGES_REQUESTED** [review]
Thanks for the changes! Just a few nitpicks left ;) Looks very good to me!

File: src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java (hunk 10 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/common/offset/EndOffsetReader.java]: Agree, I fixed it with the try-with-resources syntax .

File: src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java (hunk 11 above)
Comment (WesselVS) [reply src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java]: Removed

**Review by itadventurer — APPROVED** [review]
Cool! Looks good to me
